upf/plan_validator.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2021 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+from typing import Dict
+import upf.environment
+from upf.simplifier import Simplifier
+from upf.substituter import Substituter
+from upf.fnode import FNode
+from upf.expression import Expression
+
+
+class PlanValidator(object):
+    """Performs plan validation."""
+    def __init__(self, env: 'upf.environment.Environment'):
+        self._env = env
+        self.manager = env.expression_manager
+        self.type_checker = env.type_checker
+        self._simplifier = Simplifier(self._env)
+
+        self._substituter = Substituter(self._env)
+
+    def is_valid_plan(self, problem, plan) -> bool:
+        assignments = problem.initial_values()
+        for ai in plan.actions():
+            new_assignments: Dict[Expression, Expression] = {}
+            for ap, oe in zip(ai.action().parameters(), ai.parameters()):
+                assignments[ap] = oe
+            for p in ai.action().preconditions():
+                ps = self._subs_simplify(p, assignments)
+                if not (ps.is_bool_constant() and ps.bool_constant_value()):
+                    return False
```

Comment:
This is correct, but we can maybe give out additional information like a string stating at which step and for which precondition the plan failed.